### PR TITLE
[arm64] JIT: Redundant zero/sign extensions after ldrX/ldrsX

### LINKED ISSUE #35254: ARM64: Redundant movs done for zero extend the register
Redundant movs are done to zero the upper 32-bit of register, but if they are done after `ldr`, we can eliminate them.

```asm
B8736822          ldr     w2, [x1, x19]
2A0203E2          mov     w2, w2   ; can be eliminated
```
In framework libraries code, found 1098 such instructions that can eliminated. 

Details : 
[redundant-mov-3.txt](https://github.com/dotnet/runtime/files/4512202/redundant-mov-3.txt)

category:cq
theme:optimization
skill-level:intermediate
cost:small

### PR DESCRIPTION
**Pull request by EgorBo**
Just a quick peephole opt to drop redundant zero extensions after ldr, seems to have some nice asm diffs: up to `-177648 bytes`: https://dev.azure.com/dnceng/public/_build/results?buildId=1508477&view=ms.vss-build-web.run-extensions-tab

Minimal repro:
```csharp
static ulong Foo(ref uint p) => p;
```
Codegen diff:
```diff
; Method Prog:Foo(byref):long
G_M41239_IG01:
            stp     fp, lr, [sp,#-16]!
            mov     fp, sp
G_M41239_IG02:
            ldr     w0, [x0]
-           mov     w0, w0
G_M41239_IG03:
            ldp     fp, lr, [sp],#16
            ret     lr
-; Total bytes of code: 24
+; Total bytes of code: 20
```

Closes https://github.com/dotnet/runtime/issues/35254

### CODE REVIEW

File: src/coreclr/jit/emitarm64.cpp [hunk 1]
```diff
@@ -15629,24 +15638,58 @@ bool emitter::IsRedundantMov(instruction ins, emitAttr size, regNumber dst, regN
...
+                        JITDUMP("\n -- suppressing sxtw and chaning previous ldr to ldrsw\n");
+                    }
+                    else
+                    {
+                        JITDUMP("\n -- suppressing sxtw because ldrs already cleared upper 4 bytes\n");
+                    }
+                    return true;
+                }
+            }
+        }
```
Comment (EgorBo) [inline src/coreclr/jit/emitarm64.cpp:15686]: it's done in the emitter just to estimate impact and check tests so no review is required at this point

**Comment by EgorBo** [comment]
@dotnet/jit-contrib PTAL, very simple change with nice diffs (up to `-44376`)

File: src/coreclr/jit/lowerarmarch.cpp [hunk 2]
```diff
@@ -501,16 +500,6 @@ void Lowering::LowerCast(GenTree* tree)
...
-    if (tmpType != TYP_UNDEF)
-    {
-        GenTree* tmp = comp->gtNewCastNode(tmpType, op1, tree->IsUnsigned(), tmpType);
-        tmp->gtFlags |= (tree->gtFlags & (GTF_OVERFLOW | GTF_EXCEPT));
-
-        tree->gtFlags &= ~GTF_UNSIGNED;
-        tree->AsOp()->gtOp1 = tmp;
-        BlockRange().InsertAfter(op1, tmp);
-    }
-
```
Comment (EgorBo) [inline src/coreclr/jit/lowerarmarch.cpp:513]: dead code

**Comment by BruceForstall** [comment]
1. Please update the `IsRedundantMov` function header comment to cover this new case.
2. Did you look at why we're generating the extraneous `mov` in the first place? I.e., is it easy/possible to simply stop generating it instead of cleaning it up later?

**Comment by EgorBo** [comment]
> Did you look at why we're generating the extraneous mov in the first place? I.e., is it easy/possible to simply stop generating it instead of cleaning it up later?

Yes it's possible via a sort of an invalid JIT IR:
```
N004 (  5,  5) [000003] ---XG-------              *  RETURN    long   $280
N003 (  4,  4) [000002] ---XG----U--              \--*  CAST      long <- ulong <- uint <l:$241, c:$243>
N002 (  3,  2) [000001] *--XG-------                 \--*  IND       int    <l:$201, c:$200>
N001 (  1,  1) [000000] ------------                    \--*  LCL_VAR   byref  V00 arg0         u:1 (last use) $80
```
```
N004 (  5,  5) [000003] ---XG-------              *  RETURN    long   $280
N002 (  3,  2) [000001] *--XG-------               \--*  IND       int    <l:$201, c:$200>
N001 (  1,  1) [000000] ------------                  \--*  LCL_VAR   byref  V00 arg0         u:1 (last use) $80
```
but I'd keep it in the emitter (it's a few lines of code anyway) rather than producing sub-valid IR in lower. It's just like `ldr` cleared upper 4 bytes for us unintentionally 🙂

**Comment by EgorBo** [comment]
> Please update the IsRedundantMov function header comment to cover this new case.

Done!

**Review by kunalspathak — APPROVED** [review]
LGTM

**Pull request merged**